targets/TARGET_Cypress/TARGET_PSOC6/analogout_api.c

```diff
@@ -0,0 +1,185 @@
+{
+    if (!ctdac_initialized) {
+        uint32_t dac_clock_divider = CY_INVALID_DIVIDER;
+
+        ctdac_initialized = true;
+        // Allocate and setup clock.
+        dac_clock_divider = cy_clk_allocate_divider(CY_SYSCLK_DIV_8_BIT);
+        if (dac_clock_divider == CY_INVALID_DIVIDER) {
+            error("CTDAC clock divider allocation failed.");
+            return;
+        }
+        Cy_SysClk_PeriphSetDivider(CY_SYSCLK_DIV_8_BIT,
+                                   dac_clock_divider,
+                                   ((CY_CLK_PERICLK_FREQ_HZ + CTDAC_BASE_CLOCK_HZ / 2) / CTDAC_BASE_CLOCK_HZ) - 1);
+        Cy_SysClk_PeriphEnableDivider(CY_SYSCLK_DIV_8_BIT, dac_clock_divider);
+        Cy_SysClk_PeriphAssignDivider(obj->clock, CY_SYSCLK_DIV_8_BIT, dac_clock_divider);
+
+        Cy_CTDAC_Init(obj->base, &ctdac_config);
+        Cy_CTDAC_Enable(obj->base);
+    }
+}
+
+
+/** Initialize the analogout peripheral
+ *
+ * Configures the pin used by analogout.
+ * @param obj The analogout object to initialize
+ * @param pin The analogout pin name
+ */
+void analogout_init(dac_t *obj, PinName pin)
+{
+    uint32_t    dac = 0;
+    uint32_t    dac_function = 0;
+
+    MBED_ASSERT(obj);
+    MBED_ASSERT(pin != (PinName)NC);
+
+
+    dac = pinmap_peripheral(pin, PinMap_DAC);
+    if (dac != (uint32_t)NC) {
+        if (cy_reserve_io_pin(pin)) {
+            error("ANALOG OUT pin reservation conflict.");
+        }
+        obj->base = (CTDAC_Type*)CY_PERIPHERAL_BASE(dac);
+        obj->pin = pin;
+
+        // Configure clock.
+        dac_function = pinmap_function(pin, PinMap_DAC);
+        obj->clock = CY_PIN_CLOCK(dac_function);
+        pin_function(pin, dac_function);
+        ctdac_init(obj);
+    } else {
+        error("ANALOG OUT pinout mismatch.");
+    }
+}
+
+/** Release the analogout object
+ *
+ * Note: This is not currently used in the mbed-drivers
+ * @param obj The analogout object
+ */
+void analogout_free(dac_t *obj)
+{
+
+}
+
+/** Set the output voltage, specified as a percentage (float)
+ *
+ * @param obj The analogout object
+ * @param value The floating-point output voltage to be set
+ */
+void analogout_write(dac_t *obj, float value)
+{
+    uint32_t val = 0;
+
+    if (value > 1.0) {
+        val = CTDAC_MAX_VALUE;
+    } else if (value > 0.0) {
+        val = value * CTDAC_MAX_VALUE;
+    }
+    Cy_CTDAC_SetValueBuffered(obj->base, val);
+}
+
+/** Set the output voltage, specified as unsigned 16-bit
+ *
+ * @param obj The analogout object
+ * @param value The unsigned 16-bit output voltage to be set
+ */
+void analogout_write_u16(dac_t *obj, uint16_t value)
+{
+    uint32_t val = 0;
+
+    val = (value >> (16 - CTDAC_NUM_BITS)); // Convert from 16-bit range.
+
+    Cy_CTDAC_SetValueBuffered(obj->base, val);
+}
+
+/** Read the current voltage value on the pin
+ *
+ * @param obj The analogout object
+ * @return A floating-point value representing the current voltage on the pin,
+ *     measured as a percentage
+ */
```

Comment:
Why are we having doxygen comments here (copy paste from header file, if we improve documentation in the analogout header file, this might be outdated and hard to maintain if each target implementation does it). 

This is valid for any HAL implementation ( I can see it in every _api HAL code file).